kubernetes/utils/create_from_yaml.py

```diff
@@ -54,35 +55,46 @@ def create_from_yaml(
         processing of the request.
         Valid values are: - All: all dry run stages will be processed
     """
+    if path.exists(yaml_file):
```

Comment:
If you want it to handle a dict type input, you could check it here.

Reply:
I'd rather be explicit and have a second function do this instead of having on function to do all.
I added yaml string to load_from_yaml in order not to break previous code.

Reply:
having the behavior of the function depend on the state of the filesystem is not good
the function should always decide based on the type of the input whether it will try to treat it as a file path or something else.
when using this string is data or a path this a bit difficult to do as they are the same type.
but for a file-like interface one option would be to check for a read function and else treat it as a file path
or consider strings and path objects as paths and everything else as file-like

Reply:
@juliantaylor, basically I agree with you. I don't like Java style function with the same name but different signature. Which is also what we have here, yaml_file is two things and is completely not obvious from the function name that it expects a file. 
As said before, I'd like to have the function renamed to `create_from_file` which is explicit, but also breaks old code. So I understand why this won't be accepted. 
As a compromise, we could change the argument name from  `yaml_file` to `yaml_or_file` which is a bit more verbose, but can give a good first hint what the function expects.
@micw523 what do you think?

Reply:
my main concern here is more that the behavior of the function depends on the state of the filesystem
also os.path will not work on a file descriptor/stream which is the primary way of how I would like the function to work

```
os.path.exists(io.StringIO())
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-d85685824f54> in <module>()
----> 1 os.path.exists(io.StringIO())

/usr/lib/python3.7/genericpath.py in exists(path)
     17     """Test whether a path exists.  Returns False for broken symbolic links"""
     18     try:
---> 19         os.stat(path)
     20     except OSError:
     21         return False

TypeError: stat: path should be string, bytes, os.PathLike or integer, not _io.StringIO
```

Reply:
This PR mostly tries to fix the issue with adding the ability to create items from dictionaries. IMHO, the function `create_from_yaml` should not read a file from the file system, this should be the responsibility of the user. But since we had to keep it backwards compatible, I chose this path.
Can you make a concrete suggestion, other wise I would like to resolve this discussion and have this PR merged.

Reply:
I have made two suggestions already:
- only assume a path if it is a pathlike object
- duck typing by trying read and fallback to open

indeed the other function allows the user to solve the problem already and this functions interface can be left as is.
But it isn't as is, this whole if statement can be removed and replaced with `with open: yaml.safe_load_all()`

Reply:
Julian, I am sorry, your comment is still unclear to me. Please, can you create a gist with your suggestion of how to implement this function?

Reply:
as you have added the other function it is probably best to just leave this function as is and modify the `create_from_dict` to not catch its exception as it is very unusual api for a function that can only throw one exception to return it instead of just raising it.

This function then handles the compound exception:

```
def create_from_yaml(yaml_file, ...):
    # Load all documents from a single YAML file
    with open(yaml_file) as f:
        yml_document_all = yaml.safe_load_all(f)

    fail_exceptions = []

    for yml_document in yml_document_all:
        try:
            create_from_dict(k8s_client, yml_document, verbose, **kwargs)
        except client.rest.ApiException as api_exception:
            fail_exceptions.append(api_exception)

    if fail_exceptions:
        raise FailToCreateError(fail_exceptions)

Reply:
oh then one is missing the support for list objects ...
why does it support list objects anyway if you cannot apply them as one object?


Ok in that case I would suggest to change the `create_from_dict` function to only support single objects.
the `create_from_yaml` then uses that when it is not a list object, and implements handling of the list objects itself.
That way `create_from_dict` is a sane function with a reasonable interace users can add their special cases on top and the odd `create_from_yaml` deprecated

Reply:
the implementation would the look approximately like this:
(with the list loop removed from the `create_from_dict`, possibly with a good error message in case someone tries to apply such an object)

```python
def create_from_yaml(yaml_file, ...):
    # Load all documents from a single YAML file
    with open(yaml_file) as f:
        yml_document_all = yaml.safe_load_all(f)

    fail_exceptions = []

    for data in yml_document_all:
        if "List" in data["kind"]:
          # Could be "List" or "Pod/Service/...List"
          # This is a list type. iterate within its items
          kind = data["kind"].replace("List", "") 
          for yml_document in data["items"]:
              # Mitigate cases when server returns a xxxList object
              # See kubernetes-client/python#586
              if kind is not "": 
                  yml_document["apiVersion"] = data["apiVersion"]
                  yml_document["kind"] = kind
              try:
                  create_from_dict(k8s_client, yml_document, verbose, **kwargs)
              except client.rest.ApiException as api_exception:
                  fail_exceptions.append(api_exception)
        else:
            try:
                create_from_dict(k8s_client, data, verbose, **kwargs)
            except client.rest.ApiException as api_exception:
                fail_exceptions.append(api_exception)

    if fail_exceptions:
        raise FailToCreateError(fail_exceptions)

Reply:
As this has been going on for a while I should probably clarify what I as a library user actually want:

I have a kubernetes manifest of some basic object type (represented in some standard way python, so dictionary or a model object) which I want to apply to the cluster.
When applying I do not want to have to care about which api function is the correct one to apply, I want to call a simple function to apply this single object to the cluster or fail using standard language methods (= throw an exception)
The function should figure out which api function to use to apply it (based on its kind) and whether it needs to create or patch the object in.
I want a very basic building block which is currently missing.
The `create_from_dict` you added is very close to what I want, simplify it further by removing support for List objects and it is perfect.

Reading something from a file, apply multiple objects, handling application errors, etc. can all be implemented using the building block.

Reply:
Thank for you for adding the code example. 
I am still not convinced your suggestions are better. 
To begin with, I don't like that `create_from_dict` can't use a `List` type. It's a valid k8s object, and it is needed. 
Further, if we adopt your implementation we use `create_from_yaml` to create from file, which IMHO is a misleading name. My desire was to first add the ability to create from from string, and in a following PR split `create_from_yaml` to `create_from_yaml` and `create_from_file`. IMHO, this is much easier for the users to understand what each of this functions does (even if it is at cost of having another function).
Can we agree on leaving it like this, and let future changes to deal with it?

Reply:
`List` is not a kubernetes object type in the same sense as the non lists are, it is a wrapping structure to return multiple objects in.
There is no kubernetes api to apply a `List` object in a single step so the lowest level apply function does not need to support it.

`create_from_yaml` is already a misleading name, it was always `create_from_yaml_file` but I do not particularly care for about that function.

What I care about is a proper api for the `create_from_dict` and that not the same mistakes are made that were done when `create_from_yaml` was added by baking in too much usecase specific magic.
My remaining issue with that function is only the error handling, it should not return an exception for the user to raise.
It should do its job or raise an exception, either by only supporting single object application or if you have a good usecase to require support for `Lists` raise  the `FailedToCreateError` compound exception.

Reply:
OK, I accept this, and now `create_from_dict` throws the exception instead of  `create_from_yaml`. I hope that concludes it.